### DISCUSSION TITLE
Auto-aliasing issues when using aggregate functions in Postgres

### DISCUSSION THREAD

**Discussion by ghost** [sqlalchemy/sqlalchemy#7775] [Usage Questions] [answered]
Hey all, I've been trying to solve this issue for a week now and had no success on stackoverflow and reddit, so I hope that someone here can help me out since this is a real blocker.

So I have the following code:

```python
from sqlalchemy import Column, ForeignKey, Integer
from sqlalchemy.orm import relationship

Base = declarative_base()


class Foo(Base):
    __tablename__ = "foo"
    id = Column(Integer, primary_key=True)
    bars = relationship("Bar", uselist=True, secondary="foo_bar")
    baz_id = Column(ForeignKey("baz.id"))
    baz = relationship("Baz", back_populates="foos", lazy="joined")


class Bar(Base):
    __tablename__ = "bar"
    id = Column(Integer, primary_key=True)


class Baz(Base):
    __tablename__ = "baz"
    id = Column(Integer, primary_key=True)
    foos = relationship(Foo, uselist=True)


class FooBar(Base):
    __tablename__ = "foo_bar"
    foo_id = Column(ForeignKey(Foo.id), primary_key=True)
    bar_id = Column(ForeignKey(Bar.id), primary_key=True)

```

What I want to do is select every `Foo` with 2 `Bar`s. For that I'm using the following code:

```python
stmt = (
    select(Foo)
    .join(FooBar, FooBar.foo_id == Foo.id)
    .group_by(Foo.id)
    .having(func.count(FooBar.foo_id) == 2)
)
```

Which, as far as I know, should work fine but produces the following error when I'm trying to run it:

```
column "baz_1.id" must appear in the GROUP BY clause or be used in an aggregate function
```

Looking at the generated SQL I can see where the error comes from:

```SQL
SELECT foo.id, foo.baz_id, baz_1.id AS id_1 
FROM foo JOIN foo_bar ON foo_bar.foo_id = foo.id 
LEFT OUTER JOIN baz AS baz_1 ON baz_1.id = foo.baz_id GROUP BY foo.id 
HAVING count(foo_bar.foo_id) = :count_1
```

The issue with this is that I wasn't able to come up with a way to include `baz_1.id` in the `GROUP BY` clause. I've also tried explicitly creating an alias for `Bar`, but this only resulted in there being two aliases and still the same error.

Does anyone here have an idea / hint where to look for solving this?
Or maybe even a completely different approach of getting the desired results which avoids this issue altogether.

<details>
    <summary>Full code to reproduce</summary>

```python
from sqlalchemy import Column, ForeignKey, Integer, create_engine, func, select
from sqlalchemy.orm import declarative_base, relationship, sessionmaker

Base = declarative_base()
engine = create_engine(
    "postgresql+psycopg2://***:***@127.0.0.1:5432/***", future=True
)

Session = sessionmaker(engine)


class Foo(Base):
    __tablename__ = "foo"
    id = Column(Integer, primary_key=True)
    bars = relationship("Bar", uselist=True, secondary="foo_bar")
    baz_id = Column(ForeignKey("baz.id"))
    baz = relationship("Baz", back_populates="foos", lazy="joined")


class Bar(Base):
    __tablename__ = "bar"
    id = Column(Integer, primary_key=True)


class Baz(Base):
    __tablename__ = "baz"
    id = Column(Integer, primary_key=True)
    foos = relationship(Foo, uselist=True)


class FooBar(Base):
    __tablename__ = "foo_bar"
    foo_id = Column(ForeignKey(Foo.id), primary_key=True)
    bar_id = Column(ForeignKey(Bar.id), primary_key=True)


Base.metadata.create_all(engine)

stmt = (
    select(Foo)
    .join(FooBar, FooBar.foo_id == Foo.id)
    .group_by(Foo.id)
    .having(func.count(FooBar.foo_id) == 2)
)

Session().execute(stmt)
```
</details>

**Comment by zzzeek** [comment, marked as answer]
hey there -

as you are using group_by(), you don't have the option to use joinedload() here (this is in your Foo.baz relationship as lazy="joined"), as that will introduce additional columns that will be rejected by the database as they are not in the GROUP BY clause.

Option 1 - convert joinedload() to contains_eager() so that you can render the joinedload explicitly (note modified group_by clause):

```py
from sqlalchemy.orm import contains_eager

stmt = (
    select(Foo)
    .outerjoin(Foo.baz).options(contains_eager(Foo.baz))
    .join(FooBar, FooBar.foo_id == Foo.id)
    .group_by(Foo.id, Baz.id)
    .having(func.count(FooBar.foo_id) == 2)
)
```


Option 2:  disable joinedload with lazyload:

```py
from sqlalchemy.orm import lazyload

stmt = (
    select(Foo)
    .options(lazyload(Foo.baz))
    .join(FooBar, FooBar.foo_id == Foo.id)
    .group_by(Foo.id)
    .having(func.count(FooBar.foo_id) == 2)
)
```


Option 3: eager load baz by another means, including selectinload (recommended, very efficient, doesn't get in the way of anything else), or immediateload (like a lazy load but happens immediately):

```py

from sqlalchemy.orm import selectinload
stmt = (
    select(Foo)
    .options(selectinload(Foo.baz))
    .join(FooBar, FooBar.foo_id == Foo.id)
    .group_by(Foo.id)
    .having(func.count(FooBar.foo_id) == 2)
)
```

great MCVE I can answer very quickly with a runnable test, thanks!

**Reply by ghost** [reply]
Thanks a lot!, you've saved me! 

Just fro clarification, because I wasn't able to find anything about this in the documentation, is this behavior explained somewhere there or is it something you "just have to know"?

**Reply by zzzeek** [reply]
well it's a combination of several things that someone would know, once they see the error message.  I certainly was not aware of this particular combination but when I saw the error message, saw that it refers to additional columns that I know are added by joinedload, I know that GROUP BY is extremely picky about these kinds of things, so it was clear what was happening.   I think the most important thing is that when developing you spend a lot of time with echo=True turned on so that it can be seen what SQL is being emitted for what operations.   in 2.0 I will add some additional log output for these loads too.
